src/ponyfills/objects.js

```diff
@@ -0,0 +1,176 @@
+import deepmerge from 'deepmerge';
+
+export const deepMerge = deepmerge;
+export function deepClone(obj) {
+  return JSON.parse(JSON.stringify(obj));
+}
+
+/**
+ * Check if a variable is an Object. This function considers Null and Array to not be Objects.
+ *
+ * @param variable The variable that is to tested
+ * @returns {boolean} True if the variable is an object
+ */
+export function isObject(variable) {
+  return typeof variable === 'object' && variable !== null && !Array.isArray(variable);
+}
+
+/**
+ * Convert a hierarchical map into a flattened map
+ *
+ * @param {Object} map A map with hierarchical keys
+ * @returns {Object} A map with hierarchical keys flattened
+ */
+export function flatten(map, filter) {
+  function recurse(current, parent_key) {
+    let items = {};
+    Object.keys(current).filter(filter || function() { return true; }).forEach(function(k) {
+      let v = current[k];
+      let new_key = parent_key ? (parent_key + '.' + k) : k;
+      if (isObject(v)) {
+        items = assign(items, recurse(current[k], new_key));
```

Comment:
So GraphQL has this concept of scalars, leaf nodes where traversals stop. Any reason to treat some object nodes as atoms rather than traversing deeper?

Reply:
We have that concept in the genetics as well, but it's not really needed here because you're not performing any write or delete operations.